### PR TITLE
Separate no-op versus one-shot executions

### PR DESCRIPTION
**Pull request by aomarks**
Previously, we had a single `ScriptExecution` class which handled all scripts. However, some scripts are significantly simpler than others. Namely, we now distinguish between "no-op" and "one-shot" scripts.

A "no-op" script is one that has no command defined. It just has some files and/or dependencies, and is mostly just a pass-through for fingerprints. There is no sense in writing fingerprint files, doing caching, acquiring locks, etc. for these scripts, so we no longer do that.

A "one-shot" script is a script with an actual command. It's called "one-shot" to distinguish it from an upcoming 3rd kind of script called a [service](https://github.com/google/wireit/issues/33).

**Note to reviewer**: The `OneShotExecution` class is almost identical to what used to be called `ScriptExecution`, but with some methods moved into a new `BaseExecution` class. The `NoOpExecution` class is where the meaningful changes are, because a lot of logic that used to run around reading/writing to the `.wireit` directory no longer does.

### CODE REVIEW

File: src/execution/base.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+      );
+    } else {
+      fileHashes = [];
+    }
+
+    const cacheable =
+      // If command is undefined, then it's always safe to be cached, because
+      // the script isn't going to do anything anyway. In these cases, the
+      // fingerprint is essentially just the fingerprint of the dependencies.
+      this.script.command === undefined ||
```
[the remark this answers is not in the document]
Comment (aomarks) [reply src/execution/base.ts]: Good catch! Rewrote the comment.

File: src/executor.ts [hunk 2]
```diff
@@ -153,798 +125,31 @@ export class Executor {
...
-          const hash = createHash('sha256');
-          for await (const chunk of createReadStream(absolutePath)) {
-            hash.update(chunk as Buffer);
-          }
-          return [file.path, hash.digest('hex') as Sha256HexDigest];
-        })
+  #executeAccordingToKind(script: ScriptConfig): Promise<ExecutionResult> {
+    if (script.command === undefined) {
+      return NoOpExecution.execute(
+        // Unfortunately TypeScript does not narrow this type automatically.
```
[the remark this answers is not in the document]
Comment (aomarks) [reply src/executor.ts]: Oh right, great idea! Refactored as a union, and narrowing does indeed now work. Thanks!

**Review by AndrewJakubowicz — APPROVED** [review]
Yay for optimization!

File: CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 - The internal `.wireit/*/state` file was renamed to `.wireit/*/fingerprint`.
   Should have no effect.
 
+- If a script does not define a `"command"`, then fingerprints, lock files, and
+  cache entries are no longer written to the `.wireit` directory. This should
+  have no meaningful effect.
```
[the remark this answers is not in the document]
Comment (aomarks) [reply CHANGELOG.md]: Done, changed to:

```
- If a script does not define a `"command"`, then fingerprints, lock files, and
  cache entries are no longer written to the `.wireit` directory. This change
  should have no user-facing effect apart from a very minor performance
  improvement.
```

File: src/executor.ts [hunk 4]
```diff
@@ -130,4 +125,31 @@ export class Executor {
     }
     return promise;
   }
+
+  /**
+   *
+   */
```
[the remark this answers is not in the document]
Comment (aomarks) [reply src/executor.ts]: Removed

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply CHANGELOG.md]: Perfect!

**Pull request merged**